### PR TITLE
Add timeouts

### PR DESCRIPTION
**Pull request by triarius**
This will stop the retry after a period of time. However, each callback will still execute to completion. This means, if there is long-running code within the callback, the timeout may be exceeded.

The interaction with `WithMaxAttempts` is that the first to occur of the timeout or max attempts will halt the retry.

### CODE REVIEW

File: retrier.go [hunk 1]
```diff
@@ -86,6 +87,14 @@ func WithStrategy(strategy Strategy, strategyType string) retrierOpt {
 	}
 }
 
+// WithTimeout TODO: write more
+func WithTimeout(timeout time.Duration) retrierOpt {
+	return func(r *Retrier) {
+		r.timeout = &timeout
+		r.forever = true
```
Comment (moskyb) [inline retrier.go]: given that the meaning of the `forever` field of the struct is being muddied a little bit here, i wonder if it makes sense to remove the `forever` field entirely and replace it with `maxAttempts = math.MaxInt64` or something?

but then that would also introduce order dependency to the options functions... ie 
```Go
roko.NewRetrier(
  roko.WithMaxAttempts(100),
  roko.WithTimeout(5 * time.Minute),
)
```
will have different behaviour to 
```Go
roko.NewRetrier(
  roko.WithTimeout(5 * time.Minute),
  roko.WithMaxAttempts(100),
)
```

... but i suppose that's already a problem? i'm not sure what the action on this review is.

Comment (moskyb) [reply retrier.go]: i also think there's a valid use case for wanting both a timeout _and_ a max attempt count though, but the ergonomics of it are a little bit tricky - there's an implicit boolean there of either `retry until (attempt_count > max_attempts) AND (elapsed_time > timeout)` or `retry until (attempt_count > max_attempts) OR (elapsed_time > timeout)`

i wonder if it's just easier to leave this as explicitly undefined behaviour and get to it later?

Comment (moskyb) [reply retrier.go]: we could get really crazy with it and have something like:
```Go
maxAttemptsCriterion := func(r *roko.Retrier) bool {
  return r.AttemptCount > 15
}

timeoutCriterion := func(r *roko.Retrier) bool {
  return r.ElapsedTime > 20 * time.Minute
}

roko.NewRetrier(
  roko.WithGiveUpCriteria(
    roko.And(
      roko.Or(maxAttemptsCriterion, timeoutCriterion),
      someOtherCriterion,
  )
)        
```

and allow people to define their own give up criteria, but that seems like a pretty narrow use case, and probably defeats this library's purpose of being "an easy to use [...] retry library for go"

anyway, i've rabbitholed myself. i don't think that there's any action that we have to take right away on how different give-up criteria work (in the current implementation) right now

Comment (triarius) [reply retrier.go]: I think addressing max attempt is necessary. I think callers would expect all the conditions to be implicitly `AND`ed, so whichever is exceeded first halts the retry.

Comment (moskyb) [reply retrier.go]: > I think callers would expect all the conditions to be implicitly `AND`ed

yep, i think that's an extremely reasonable assumption to make

**Comment by triarius** [comment]
So compared to passing a context with timeout to `DoWithContext` directly, this has the property that the timeout is not available inside the callback.

In particular, in the following code
```go
ctx := context.Background()
req, _ := http.NewRequestWithContext(ctx, "GET", "https://httpstat.us/500", nil)
err := NewRetrier(
	WithStrategy(Constant(time.Second)),
	WithTimeout(timeout),
).DoWithContext(ctx, func(*Retrier) error {
	resp, err := client.Do(req)
	if err != nil {
		return err
	}
	if resp.StatusCode != http.StatusOK {
		return errors.New("not OK")
	}
	return nil
})
```
The HTTP calls will always wait for a response from the server and are not interrupted by the roko timeout. On the other hand, if the timeout were set outside the retrier, passing it into `NewRequestWithContext` would cause the http client to abort the connection.


Is this what a caller expects? Possibly yes.

**Pull request closed without merging**